File: scala/CardRaytracer/src/main/scala/Vector.scala
import java.lang.Math._

class Vector(val x: Double,val y: Double,val z: Double) {

  def this() {
    this(0.0, 0.0, 0.0)
  }

  def +(r: Vector): Vector = {
    new Vector(x + r.x, y + r.y, z + r.z)
  }

  def *(r: Double): Vector = {
    new Vector(x * r, y * r, z * r)
  }

  def %(r: Vector): Double = {
    x * r.x + y * r.y + z * r.z
  }

  def ^(r: Vector): Vector = {
    new Vector(y * r.z - z * r.y,
      z * r.x - x * r.z,
      x * r.y - y * r.x)
  }

  def unary_! : Vector = this * (1.0/sqrt(this % this))

  override def toString() : String = s"{$x,$y,$z}"

}

File: java-mt/src/main/java/mayton/Vector.java
package mayton;

import javax.annotation.concurrent.Immutable;

import static java.lang.Math.sqrt;

/**
 * Represents triple of floating-point numbers.
 */
@Immutable
public final class Vector {

    public final double x;
    public final double y;
    public final double z;

    public Vector(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector sum(Vector r) {
        return new Vector(x + r.x, y + r.y, z + r.z);
    }

    public Vector prod(double r) {
        return new Vector(x * r, y * r, z * r);
    }

    public double sprod(Vector r) {
        return x * r.x + y * r.y + z * r.z;
    }

    public Vector vprod(Vector r) {
        return new Vector(y * r.z - z * r.y,
                z * r.x - x * r.z,
                x * r.y - y * r.x);
    }

    public Vector norm() {
        return this.prod(1.0 / sqrt(this.sprod(this)));
    }

}

